pkg/apis/databases/v1alpha4/connection.go

```diff
@@ -0,0 +1,224 @@
+/*
+Copyright 2019 Replicated, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha4
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"text/template"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+// GetConnection will return a valid connection string for the database. This
+// is compatible with any way that the uri was set.
+// TODO refactor this to be shorter, simpler and more testable
+func (d Database) GetConnection(ctx context.Context) (string, string, error) {
```

Comment:
is this new code? or is this copied from alpha3?

Reply:
its new.  this was spread around the codebase, and now in a single place